File: Trees/q18_numberOfMeanNodes.cpp

int number_of_mean_nodes(TreeNode
                             *a)
{
    if (a == nullptr)
    {
        return 0;
    }

    int count = 0;
    if (a->left != nullptr && a->right != nullptr)
    {
        int leftValue = a->left->val;
        int rightValue = a->right->val;

        // Check if the node's value is the mean of its children's values
        if (a->val == (leftValue + rightValue) / 2)
        {
            count = 1; // This is a mean node
        }
    }

    // Recur for left and right subtrees
    count += number_of_mean_nodes(a->left);
    count += number_of_mean_nodes(a->right);

    return count;
}

int numberOfMeanNodes(TreeNode
                          *root)
{
    return number_of_mean_nodes(root);
}
}

File: MT2/Trees/q18_numberOfMeanNodes.java
package Trees;

public class q18_numberOfMeanNodes {

    /*
Write a function which returns the number of mean nodes. A node is a
 mean node if its value is the mean of its left and right children’s values.
 int numberOfMeanNodes()
     */
    int numberOfMeanNodes() {
        int count = 0;
        if (left != null && right != null) {
            if (this.data == (left.data + right.data) / 2) {
                count++;
            }
        }
        if (left != null) {
            count += left.numberOfMeanNodes();
        }
        if (right != null) {
            count += right.numberOfMeanNodes();
        }
        return count;
    }

}
